sycl/include/sycl/ext/oneapi/experimental/annotated_usm/alloc_util.hpp

```diff
@@ -0,0 +1,236 @@
+};
+
+// Filter the compile-time properties from a property list
+template <typename PropertyListT> struct GetCompileTimeProperties {};
+
+template <> struct GetCompileTimeProperties<detail::empty_properties_t> {
+  using type = detail::empty_properties_t;
+};
+
+template <typename Prop>
+struct GetCompileTimeProperties<detail::properties_t<Prop>> {
+  using type =
+      std::conditional_t<detail::IsCompileTimePropertyValue<Prop>::value,
+                         detail::properties_t<Prop>,
+                         detail::empty_properties_t>;
+};
+
+template <typename Prop, typename... Props>
+struct GetCompileTimeProperties<detail::properties_t<Prop, Props...>> {
+  using filtered_this_property_t =
+      std::conditional_t<detail::IsCompileTimePropertyValue<Prop>::value,
+                         detail::properties_t<Prop>,
+                         detail::empty_properties_t>;
+  using filtered_other_properties_t =
+      typename GetCompileTimeProperties<detail::properties_t<Props...>>::type;
+  using type = detail::merged_properties_t<filtered_this_property_t,
+                                           filtered_other_properties_t>;
+};
+
+// Given the input property list of annotated USM allocation functions, generate
+// the property list for the annotated_ptr output of
+template <alloc Kind, typename PropertyListT>
+struct GetAnnotatedPtrPropertiesWithUsmKind {};
+
+// Partial specialization
+template <alloc Kind, typename... Props>
+struct GetAnnotatedPtrPropertiesWithUsmKind<Kind,
+                                            detail::properties_t<Props...>> {
+  using input_properties_t = detail::properties_t<Props...>;
+  using filtered_input_properties_t =
+      typename GetCompileTimeProperties<input_properties_t>::type;
+
+  static_assert(!HasUsmKind<input_properties_t>::value ||
+                    GetUsmKindFromPropList<input_properties_t>::value == Kind,
+                "Input property list contains conflicting USM kind.");
+
+  using type =
+      detail::merged_properties_t<filtered_input_properties_t,
+                                  decltype(properties{usm_kind<Kind>})>;
+};
+
+// Validate the template arguments for annotated USM allocation functions, which
+// consist of
+//   allocated data type: T
+//   input property list: propertyListA
+//   property list for output anntoated_ptr: propertyListB
+template <typename T, typename propertyListA, typename propertyListB>
+struct CheckTAndPropLists : std::false_type {};
+
+template <typename T, typename... PropsA, typename... PropsB>
+struct CheckTAndPropLists<T, detail::properties_t<PropsA...>,
+                          detail::properties_t<PropsB...>>
+    : std::conditional_t<
+          is_property_list<T>::value, std::false_type,
+          std::is_same<detail::properties_t<PropsB...>,
+                       typename GetCompileTimeProperties<
+                           detail::properties_t<PropsA...>>::type>> {};
```

Comment:
Can you add a comment about what this is validating and why?

Reply:
comments added